google-cloud-logging/lib/google/cloud/logging/entry.rb

```diff
@@ -82,6 +83,7 @@ def initialize
           @operation = Operation.new
           @severity = :DEFAULT
           @source_location = SourceLocation.new
+          @insert_id = SecureRandom.uuid
```

Comment:
I'm inclined not to use `SecureRandom` here, for performance reasons. Log entries need to be virtually instantaneous to create, and `SecureRandom` makes calls to get OS-provided entropy, which could block if the entropy generator is running behind. This doesn't really need to be cryptographically secure anyway.

Reply:
How random should it be? Would something like this be sufficient?

```ruby
rand(36**16).to_s(36) #=> “omyizc4969m4d3pc”
```

Reply:
Here are some options for converting a random integer of different sizes and converting to string of base 36, as well as a time-based value that joins the current time as an integer in base 36, the current time's nanoseconds in base 36, and a small random integer in base 36:

```
uuid:       70304e71-2df2-4bcc-9acd-d5d649652b4b
rand(8):    s2i0dos
rand(16):   qi9lpczfp5513sp6
rand(24):   g3on0crn0itrxkcb34vhpsm6
rand(32):   h0qolmzqjllps69sku49eyyq8o5j3oad
time-based: plhl3v-1szz74-lwaz2odz
```

Here are how those different options perform when generating one million of them:

```
              user     system      total        real
uuid:        3.016788   0.012704   3.029492 (  3.040411)
rand(8):     0.300438   0.000650   0.301088 (  0.301689)
rand(16):    0.868608   0.002544   0.871152 (  0.873399)
rand(24):    1.363830   0.004168   1.367998 (  1.372142)
rand(32):    1.614370   0.003650   1.618020 (  1.620551)
time-based:  1.565573   0.004644   1.570217 (  1.573589)
```

Where is your comfort level for uniqueness vs. performance?

Reply:
Here is the benchmark code I used:

```ruby
require "benchmark"
require "securerandom"

def random_id size
  rand(36**size).to_s(36)
end

def time_based_id size
  now = Time.now
  [
    now.to_i.to_s(36),
    now.nsec.to_s(36),
    random_id(size)
  ].join "-".freeze
end

n = 1000000
Benchmark.bm(7) do |x|
  x.report("uuid:      ") { n.times { SecureRandom.uuid } }
  x.report("rand(8):   ") { n.times { random_id(8) } }
  x.report("rand(16):  ") { n.times { random_id(16) } }
  x.report("rand(24):  ") { n.times { random_id(24) } }
  x.report("rand(32):  ") { n.times { random_id(32) } }
  x.report("time-based:") { n.times { time_based_id(8) } }
end

puts ""
puts "Example output"
puts "uuid:       #{SecureRandom.uuid}"
puts "rand(8):    #{random_id(8)}"
puts "rand(16):   #{random_id(16)}"
puts "rand(24):   #{random_id(24)}"
puts "rand(32):   #{random_id(32)}"
puts "time-based: #{time_based_id(8)}"
```

Reply:
Something quick should be sufficient. The backend uses a combination of the timestamp and insert_id to dedup inserts. Given that timestamp has microsecond resolution on macOS, and nanosecond resolution on Linux, the chances of collisions are very very small.

Reply:
This has been updated. PTAL.